packages/core/src/provider/base.ts

```diff
@@ -487,59 +495,149 @@ export class SmartAccountProvider<
   // You should implement your own middleware to override these
   // or extend this class and provider your own implemenation
   readonly dummyPaymasterDataMiddleware: AccountMiddlewareFn = async (
-    struct
+    struct,
+    overrides,
+    feeOptions
   ) => {
+    Logger.debug(
+      `[SmartAccountProvider] dummyPaymasterDataMiddleware`,
+      struct,
+      overrides,
+      feeOptions
+    );
     struct.paymasterAndData = "0x";
     return struct;
   };
 
-  readonly paymasterDataMiddleware: AccountMiddlewareFn = async (struct) => {
-    struct.paymasterAndData = "0x";
+  readonly overridePaymasterDataMiddleware: AccountMiddlewareFn = async (
+    struct,
+    overrides,
+    feeOptions
+  ) => {
+    Logger.debug(
+      `[SmartAccountProvider] overridePaymasterDataMiddleware`,
+      struct,
+      overrides,
+      feeOptions
+    );
+    struct.paymasterAndData =
+      overrides?.paymasterAndData != null ? overrides?.paymasterAndData : "0x";
     return struct;
   };
 
-  readonly gasEstimator: AccountMiddlewareFn = async (struct) => {
-    const request = deepHexlify(await resolveProperties(struct));
-    const estimates = await this.rpcClient.estimateUserOperationGas(
-      request,
-      this.getEntryPointAddress()
+  readonly paymasterDataMiddleware: AccountMiddlewareFn = async (
+    struct,
+    overrides,
+    feeOptions
+  ) => {
+    Logger.debug(
+      `[SmartAccountProvider] paymasterDataMiddleware`,
+      struct,
+      overrides,
+      feeOptions
     );
-
-    struct.callGasLimit = estimates.callGasLimit;
-    struct.verificationGasLimit = estimates.verificationGasLimit;
-    struct.preVerificationGas = estimates.preVerificationGas;
-
+    struct.paymasterAndData = "0x";
     return struct;
   };
 
-  readonly feeDataGetter: AccountMiddlewareFn = async (struct) => {
-    const [maxPriorityFeePerGas, feeData] = await Promise.all([
-      this.rpcClient.estimateMaxPriorityFeePerGas(),
-      this.rpcClient.estimateFeesPerGas(),
-    ]);
-    if (!feeData.maxFeePerGas || !feeData.maxPriorityFeePerGas) {
-      throw new Error(
-        "feeData is missing maxFeePerGas or maxPriorityFeePerGas"
+  readonly gasEstimator: AccountMiddlewareFn = async (
+    struct,
+    overrides,
+    feeOptions
+  ) => {
+    Logger.debug(
+      `[SmartAccountProvider] gasEstimator`,
+      struct,
+      overrides,
+      feeOptions
+    );
+
+    let { callGasLimit, verificationGasLimit, preVerificationGas } =
+      overrides ?? {};
+
+    if (
+      callGasLimit === undefined ||
+      verificationGasLimit === undefined ||
+      preVerificationGas === undefined
```

Comment:
nit, we should use `== null` when doing these checks. There's some peeps that don't use typescript so there's chance they might set these incorrectly

Reply:
bump

Reply:
so weird. I thought I already changed all these... hmm